source/s3_auto_ranged_put.c

```diff
@@ -598,6 +573,106 @@ static bool s_s3_auto_ranged_put_update(
+ * that part). If checksum is set on the request and parts with checksums were uploaded before, checksum will be
+ * verified.
+ */
+static int s_skip_parts_from_stream(
+    struct aws_s3_meta_request *meta_request,
+    uint32_t num_parts_read_from_stream,
+    uint32_t skip_until_part_number) {
+
+    AWS_PRECONDITION(meta_request);
+
+    struct aws_s3_auto_ranged_put *auto_ranged_put = meta_request->impl;
+
+    AWS_PRECONDITION(skip_until_part_number <= auto_ranged_put->synced_data.total_num_parts);
+
+    if (num_parts_read_from_stream == skip_until_part_number) {
+        return AWS_ERROR_SUCCESS;
+    }
+
+    struct aws_byte_buf temp_body_buf;
+    aws_byte_buf_init(&temp_body_buf, meta_request->allocator, 0);
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_S3_META_REQUEST,
+        "id=%p: Skipping parts %d through %d",
+        (void *)meta_request,
+        num_parts_read_from_stream,
+        skip_until_part_number);
+
+    for (uint32_t part_index = num_parts_read_from_stream; part_index < skip_until_part_number; ++part_index) {
+
+        size_t request_body_size = s_compute_request_body_size(meta_request, part_index + 1);
+
+        if (temp_body_buf.capacity != request_body_size) {
+            // reinit with correct size
+            aws_byte_buf_clean_up(&temp_body_buf);
+            aws_byte_buf_init(&temp_body_buf, meta_request->allocator, request_body_size);
+        } else {
+            // reuse buffer
+            aws_byte_buf_reset(&temp_body_buf, false);
+        }
+
+        if (aws_s3_meta_request_read_body(meta_request, &temp_body_buf)) {
+            AWS_LOGF_ERROR(AWS_LS_S3_META_REQUEST, "Failed to resume upload. Input steam cannot be read.");
+            aws_byte_buf_clean_up(&temp_body_buf);
+            return AWS_ERROR_S3_SKIP_UPLOADED_PART_FAILED;
+        }
+
+        // compare skipped checksum to previously uploaded checksum
+        if (meta_request->checksum_algorithm != AWS_SCA_NONE && auto_ranged_put->checksums_list[part_index].len > 0) {
+            struct aws_byte_buf checksum;
+            aws_byte_buf_init(
+                &checksum,
+                meta_request->allocator,
+                aws_get_digest_size_from_algorithm(meta_request->checksum_algorithm));
+            struct aws_byte_cursor body_cur = aws_byte_cursor_from_buf(&temp_body_buf);
+            if (!aws_checksum_compute(
+                    meta_request->allocator, meta_request->checksum_algorithm, &body_cur, &checksum, 0) ||
+                !aws_byte_buf_eq(&checksum, &auto_ranged_put->checksums_list[part_index])) {
+                AWS_LOGF_ERROR(
+                    AWS_LS_S3_META_REQUEST,
+                    "Failed to resume upload. Checksum for previously uploaded part does not match "
+                    "new part.");
+                aws_byte_buf_clean_up(&temp_body_buf);
+                aws_byte_buf_clean_up(&checksum);
+                return AWS_ERROR_S3_RESUMED_PART_CHECKSUM_MISMATCH;
```

Comment:
well, if `aws_checksum_compute(meta_request->allocator, meta_request->checksum_algorithm, &body_cur, &checksum, 0)` failed, we will still raise the error of `AWS_ERROR_S3_RESUMED_PART_CHECKSUM_MISMATCH`.